### PR TITLE
feat: customer ticket portal

### PR DESCRIPTION
**Pull request by atiq-ur**
1. Add Helpdesk settings on Thrivedesk setting option
2. Add Conversation Shortcode
3. Add Conversation list page with pagination
4. Add Conversation detail page
5. Add reply option on detail page (using TinyMCE editor)
6. Add new ticket modal with dynamic search option

UI  - [Video link](https://www.loom.com/share/52dbe5d7ae044dc68381ba1336c64cc9)

### CODE REVIEW

File: includes/views/shortcode/modal.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		<div class="relative bg-white rounded-lg shadow dark:bg-gray-700">
+			<!-- Modal header -->
+			<div class="flex justify-between items-start p-4 rounded-t border-b dark:border-gray-600">
+				<h3 class="text-xl font-semibold text-gray-900 dark:text-white">
+					Create new ticket
+				</h3>
+				<button type="button" id="close_modal" class="text-gray-400 bg-transparent hover:bg-gray-200
+                hover:text-gray-900
+                rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white" data-modal-toggle="tdConversationModal">
+					<svg aria-hidden="true" class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z" clip-rule="evenodd"></path></svg>
```
Comment (alaminfirdows) [inline includes/views/shortcode/modal.php:24]: move it to file

File: src/Conversations/Conversation.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+	/**
+	 * redirect to the conversation page
+	 * if conversation id then redirect to the conversation details page
+	 *
+	 */
+	public function conversation_page() {
+		ob_start();
+		if (isset($_GET['conversation_id'])) {
+			include THRIVEDESK_DIR. '/includes/views/shortcode/conversation-details.php';
```
Comment (alaminfirdows) [inline src/Conversations/Conversation.php:87]: use `thrivedesk_view`

File: src/Conversations/Conversation.php [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		return $body['data'];
+	}
+
+	/**
+	 * validate html body of the conversation
+	 * it will help to remove style breaking issue on event body
+	 * @return false|string
+	 */
+	public static function validate_conversation_body($content){
+		$dom = new DOMDocument('1.0', 'UTF-8');
```
Comment (alaminfirdows) [inline src/Conversations/Conversation.php:163]: why we need this?

File: src/Conversations/Conversation.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		@$dom->loadHTML($content);
+		return $dom->saveHTML();
+	}
+
+	/*
+	 * admin setting
+	 * get add settings from the wp options
+	 */
+	public function getSelectedHelpdeskOptions() {
```
Comment (alaminfirdows) [inline src/Conversations/Conversation.php:172]: it should be moved to parent

File: src/Conversations/Conversation.php [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		$body     = wp_remote_retrieve_body( $response );
+		$body     = json_decode( $body, true );
+		return $body ?? [];
+	}
+
+	/**
+	 * format timestamp for the conversation
+	 * @throws \Exception
+	 */
+	public static function diffForHumans($datetime, $full = false): string {
```
Comment (alaminfirdows) [inline src/Conversations/Conversation.php:199]: move it to helper

File: src/FormProviders/FormProviderHelper.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace ThriveDesk\FormProviders;
+
+class FormProviderHelper {
```
Comment (alaminfirdows) [inline src/FormProviders/FormProviderHelper.php:5]: Do we need this?

File: src/RestRoute.php [hunk 7]
```diff
@@ -31,6 +38,7 @@ public static function instance()
     private function __construct()
     {
         add_action('rest_api_init', array($this, 'td_routes'));
+        add_action('rest_api_init', array($this, 'form_routes'));
```
Comment (alaminfirdows) [inline src/RestRoute.php:41]: what is `form_routes`?

File: thrivedesk.php [hunk 8]
```diff
@@ -110,6 +113,7 @@ private function define_constants(): void
         $this->define('THRIVEDESK_PLUGIN_ASSETS_PATH', plugin_dir_path(__FILE__) . 'assets');
         // Url with no ending /
         $this->define('THRIVEDESK_APP_URL', 'https://app.thrivedesk.com');
+        $this->define('THRIVEDESK_API_URL', 'http://api.thrivedesk.test');
```
Comment (alaminfirdows) [inline thrivedesk.php:116]: change to the prod url

**Review by alaminfirdows — CHANGES_REQUESTED** [review]

**Comment by alaminfirdows** [comment]
You should clean the codebase, organize css class names and re-indent.

File: src/FormProviders/FormProviderHelper.php (hunk 6 above)
Comment (atiq-ur) [reply src/FormProviders/FormProviderHelper.php:5]: no

File: src/RestRoute.php (hunk 7 above)
Comment (atiq-ur) [reply src/RestRoute.php:41]: should be conversation routes

File: includes/views/shortcode/conversations.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                            ?></a>
+                    </li>
+                <?php endforeach; ?>
+			</ul>
+		</div>
+	</div>
+</div>
+
+<?php
+include THRIVEDESK_DIR. '/includes/views/shortcode/modal.php';
```
[the remark this answers is not in the document]
Comment (atiq-ur) [reply includes/views/shortcode/conversations.php]: solved by adding ```thrivedesk_view```

File: src/Conversations/Conversation.php (hunk 3 above)
Comment (atiq-ur) [reply src/Conversations/Conversation.php:163]: Need to parse html body content, means if html body does not contain end tag then, it brakes the style of the conversation. That is why we catch the errors and shows the actual element.

File: src/Conversations/Conversation.php (hunk 5 above)
Comment (atiq-ur) [reply src/Conversations/Conversation.php:199]: moved

File: src/FormProviders/FormProviderHelper.php (hunk 6 above)
Comment (atiq-ur) [reply src/FormProviders/FormProviderHelper.php:5]: deleted

File: src/RestRoute.php (hunk 7 above)
Comment (atiq-ur) [reply src/RestRoute.php:41]: moved to td routes

File: includes/views/shortcode/modal.php (hunk 1 above)
Comment (atiq-ur) [reply includes/views/shortcode/modal.php:24]: moved

File: src/Conversations/Conversation.php (hunk 4 above)
Comment (atiq-ur) [reply src/Conversations/Conversation.php]: Resolved. Moved to helper function

File: thrivedesk.php (hunk 8 above)
Comment (atiq-ur) [reply thrivedesk.php:116]: prod url should be changed on deploy time

**Pull request merged**